### PR TITLE
Add custom numeric editor

### PR DESCRIPTION
**Pull request by simasjar**

### CODE REVIEW

File: packages/components/src/presentation-components/properties/NumericPropertyEditor.tsx [hunk 1]
```diff
@@ -0,0 +1,74 @@
+import { createRef, PureComponent } from 'react';
```
Comment (saskliutas) [inline packages/components/src/presentation-components/properties/NumericPropertyEditor.tsx:1]: Missing copy right header.

File: packages/components/src/presentation-components/properties/NumericPropertyEditor.tsx [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+import { StandardTypeNames } from '@itwin/appui-abstract';
```
Comment (saskliutas) [inline packages/components/src/presentation-components/properties/NumericPropertyEditor.tsx:3]: Imports sorted  different than in other files. Could you copy import sorter settings from `presentation/.vscode/settings.json` to your workspace if you are using `CoSpace` and rerun import sorter?

File: packages/components/src/presentation-components/properties/NumericPropertyEditor.tsx [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+import { NumericPropertyInput, NumericPropertyInputAttributes
+} from './NumericPropertyInput';
+
+/**
+ * Name for `NumericPropertyEditor`.
+ *
+ * @internal
+ */
+export const NumericEditorName = "numeric-editor";
```
Comment (saskliutas) [inline packages/components/src/presentation-components/properties/NumericPropertyEditor.tsx:16]:
```suggestion
export const NumericEditorName = "presentation-numeric-editor";
```

File: packages/components/src/presentation-components/properties/NumericPropertyInput.tsx [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+export const NumericPropertyInput = forwardRef<NumericPropertyInputAttributes, NumericPropertyInputProps>((props, ref) => {
+  const { onCommit, propertyRecord } = props;
+
+  const [inputValue, setInputValue] = useState<string>(() => getInputTargetFromPropertyRecord(propertyRecord) ?? "");
+
+  const divRef = useRef<HTMLInputElement>(null);
+  useImperativeHandle(
+    ref,
+    () => ({
+      getValue: () => ({ valueFormat: PropertyValueFormat.Primitive, value: isNaN(Number(inputValue)) ? undefined : Number(inputValue), inputValue }),
```
Comment (saskliutas) [inline packages/components/src/presentation-components/properties/NumericPropertyInput.tsx:28]: Shouldn't this be?
```suggestion
      getValue: () => ({ valueFormat: PropertyValueFormat.Primitive, value: isNaN(Number(inputValue)) ? undefined : Number(inputValue), displayValue: inputValue }),
```

File: packages/components/src/test/properties/NumericPropertyEditor.test.tsx [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+  record.property.typename = "number:numeric-editor";
+  return record;
+};
+
+describe("<NumericPropertyEditor />", () => {
+  beforeEach(async () => {
+    const localization = new EmptyLocalization();
+    sinon.stub(IModelApp, "initialized").get(() => true);
+    sinon.stub(IModelApp, "localization").get(() => localization);
+    await Presentation.initialize();
```
Comment (saskliutas) [inline packages/components/src/test/properties/NumericPropertyEditor.test.tsx:33]: Is this needed?

File: packages/components/src/test/properties/NumericPropertyEditor.test.tsx [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    const inputContainer = await waitFor(() => getByTestId("numeric-input"));
+
+    await user.type(inputContainer, "1");
+
+    await waitFor(() => expect(queryByDisplayValue("1")).to.not.be.null);
+    expect(spy).to.be.calledOnce;
+  });
+});
+
+describe("<NumericPropertyEditor />", () => {
```
Comment (saskliutas) [inline packages/components/src/test/properties/NumericPropertyEditor.test.tsx:62]: Why these tests are under different describe block that has same name? Looks like they should be merged together,

File: packages/components/src/test/common/ContentBuilder.test.ts [hunk 7]
```diff
@@ -21,6 +21,7 @@ import {
   createTestPropertiesContentField,
   createTestSimpleContentField,
 } from "../_helpers/Content";
+import { NumericEditorName } from "../../presentation-components";
```
Comment (saskliutas) [inline packages/components/src/test/common/ContentBuilder.test.ts:24]: Don't import from barrel file

File: packages/components/src/test/properties/NumericPropertyInput.test.tsx [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+    sinon.stub(IModelApp, "localization").get(() => localization);
+    await Presentation.initialize();
+  });
+
+  afterEach(async () => {
+    Presentation.terminate();
+    sinon.restore();
+  });
+
+  it("get value from target input reference", async () => {
```
Comment (saskliutas) [inline packages/components/src/test/properties/NumericPropertyInput.test.tsx:26]:
```suggestion
  it("renders with initial value", async () => {
```

**Review by saskliutas — CHANGES_REQUESTED** [review]

File: packages/components/src/presentation-components/properties/NumericPropertyEditor.tsx (hunk 1 above)
Comment (simasjar) [reply packages/components/src/presentation-components/properties/NumericPropertyEditor.tsx:1]: added

File: packages/components/src/presentation-components/properties/NumericPropertyEditor.tsx (hunk 2 above)
Comment (simasjar) [reply packages/components/src/presentation-components/properties/NumericPropertyEditor.tsx]: fixed?

File: packages/components/src/test/properties/NumericPropertyEditor.test.tsx (hunk 5 above)
Comment (simasjar) [reply packages/components/src/test/properties/NumericPropertyEditor.test.tsx]: removed

File: packages/components/src/test/properties/NumericPropertyInput.test.tsx (hunk 8 above)
Comment (simasjar) [reply packages/components/src/test/properties/NumericPropertyInput.test.tsx]: I think that we should change to `NumericPropertyInput`

File: packages/components/src/presentation-components/properties/NumericPropertyInput.tsx [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+    // Number("+.") and Number("-.") returns NaN, but if input is only `+.` or `-.`, we want to fire `onChange` function.
+    else if (val === "+." || val === "-.") {
+      onChange(val);
+    }
+    // Let user write scientific numbers. Number("1e") returns NaN, but we want to fire `onChange` function when input before `e` is a correct number.
+    else if (val.endsWith("e") && !isNaN(Number(val.slice(0, val.length - 1))) && val.length !== 1) {
+      onChange(val);
+    }
+    // Let user write scientific numbers. Number("1e-") returns NaN, but we want to fire `onChange` function when input before `e-` is a correct number.
+    else if (val.endsWith("e-")) {
```
Comment (saskliutas) [inline packages/components/src/presentation-components/properties/NumericPropertyInput.tsx:85]: Looks like this also should check if string before `e-` is valid number.

**Review by saskliutas — CHANGES_REQUESTED** [review]

File: packages/components/src/presentation-components/properties/NumericPropertyInput.tsx (hunk 9 above)
Comment (simasjar) [reply packages/components/src/presentation-components/properties/NumericPropertyInput.tsx:85]: Not really, there is a check if the string before `e` is a valid number, thus we don't need to check if the string before `e-` is a valid number.

File: packages/components/src/test/properties/NumericPropertyInput.test.tsx [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+    const { getByRole } = render(<NumericPropertyInput ref={ref} propertyRecord={record} onCommit={() => {}}/>);
+
+    const inputContainer = await waitFor(() => getByRole("textbox"));
+
+    await user.type(inputContainer, "-");
+
+    await waitFor(() => expect((ref.current?.getValue() as PrimitiveValue).value).to.be.undefined);
+  });
+
+  describe("integration with <NumericInput /> ", () => {
```
[the remark this answers is not in the document]
Comment (simasjar) [reply packages/components/src/test/properties/NumericPropertyInput.test.tsx]: removed

File: packages/components/src/test/properties/NumericPropertyInput.test.tsx [hunk 11]
```diff
@@ -0,0 +1,268 @@
...
+      expect(getByDisplayValue("1e-")).to.not.be.null;
+
+      await user.type(inputContainer, "5");
+      expect((ref.current?.getValue() as PrimitiveValue).value).to.be.eq(0.00001);
+    });
+  });
+});
+
+describe("<NumericInput />", () => {
+  it("renders NumericInput", () => {
```
[the remark this answers is not in the document]
Comment (simasjar) [reply packages/components/src/test/properties/NumericPropertyInput.test.tsx]: changed

**Review by saskliutas — CHANGES_REQUESTED** [review]

File: packages/components/src/test/properties/NumericPropertyInput.test.tsx [hunk 12]
```diff
@@ -0,0 +1,274 @@
...
+    const ref = createRef<NumericPropertyInputAttributes>();
+    const { getByDisplayValue, getByRole } = render(<NumericPropertyInput propertyRecord={record} ref={ref}/>);
+
+    const inputContainer = await waitFor(() => getByRole("textbox"));
+
+    await user.type(inputContainer, "-");
+    expect(getByDisplayValue("-")).to.not.be.null;
+
+    await user.type(inputContainer, "1");
+    expect((ref.current?.getValue() as PrimitiveValue).value).to.be.eq(-1);
```
[the remark this answers is not in the document]
Comment (simasjar) [reply packages/components/src/test/properties/NumericPropertyInput.test.tsx]: fixed

**Review by saskliutas — APPROVED** [review]

**Pull request merged**